### DISCUSSION TITLE
IBAMR config

### DISCUSSION THREAD

**Discussion by agiordan96** [IBAMR/IBAMR#1712] [General]
Hi all, 

I am trying to configure IBAMR (autotool) on my Mac M1. During configuration, I get "configure: error: cannot find libmesh-config". Surfing in lib mesh, the file exists, but it is found in lib mesh/1.7.1/objs-debug/contrib/bin
How to edit the configuration process and allow IBAMR to find it? I tried several ways but I failed. Same happens to the CMake config process. 

Thanks for helping out!
Andrea

**Comment by boyceg** [comment]
I believe that IBAMR's autotools-based `configure` script need to be told where libMesh is installed. It looks like you might be pointing it to the directory where you built libMesh.

**Reply by agiordan96** [reply]
Hi boyceg, thanks for the reply. I tried that one too, but it does not work either. Getting the same exact error. I am attaching here my config.log file, if you could have a look. I am following the instructions on your website step by step. 

[config.log](https://github.com/IBAMR/IBAMR/files/15391935/config.log)

**Reply by boyceg** [reply]
Please try using the directory where libMesh is installed and send the resulting config.log.

You might also consider trying [autoibamr](https://github.com/IBAMR/autoibamr), which should "just work".

**Reply by agiordan96** [reply]
Hi, that is the directory where lo mesh is installed. Why do you think otherwise, if I may?

**Reply by abarret** [reply]
If you are following the instructions on the website, it suggests installing in a directory such as `1.7.1-debug`. Following the naming conventions on the website, `libmesh-1.7.1` (as is used in the config.log you sent) would suggest the source directory. Can you confirm that the install directory is `libmesh-1.7.1`? Can you send the `config.log` of libmesh?

**Reply by agiordan96** [reply]
Hi abarret, see my answer to both of you below and thanks!

**Comment by boyceg** [comment]
> Hi, that is the directory where lo mesh is installed. Why do you think otherwise, if I may?

I believe that if `LIBMESH_INSTALL_DIR` is the libMesh installation directory, then `libmesh-config` should be present in `LIBMESH_INSTALL_DIR/bin`. If you are finding it in `contrib/bin`, then that sounds like the directory where libMesh was built, not the directory where libMesh was installed.

**Reply by agiordan96** [reply]
Hello both, thanks for your help. 
If I follow exactly what mentioned on the website, I end using a folder "1.7.1-debug" that is never created and thus cannot be found after the installation. In that case, what would the installation directory be? I thought you meant "objs-debug", and used that one during the configuration process ((config.log renamed to config1.log and attached)
[Uploading config1.log…]()
. No folder shows a "bin" subfolder.
On the other hand, creating a "1.7.1-debug" folder leads to an installation (I suppose) in that folder, but once again no bin folder is present (config.log renamed to configagain.log and attached)

**Reply by boyceg** [reply]
It seems like the `config.log` file did not get attached.

Note that the `--prefix` flag tells `configure` the directory where libMesh should be installed, and that directory should be created automatically by `make install`. Specifically, when you configure, build, and install libMesh with:
```
$ cd /path/to/libmesh-build-dir/
$ /path/to/libmesh-source-dir/configure --prefix=/path/to/libmesh-install-dir/1.7.1-debug <other flags>
$ make
$ make install
```
then `/path/to/libmesh-install-dir/1.7.1-debug` will be created by `make install`. If you need to create that directory yourself, then something has gone awry in either the configuration or build step.

**Reply by agiordan96** [reply]
Hi, that s great to know, thank you. My installation indeed gets an error, make error 1, but I thought you were not the correct person to ask to. I am trying to attach the files again!
[config1.log](https://github.com/IBAMR/IBAMR/files/15402647/config1.log)
[configagain.log](https://github.com/IBAMR/IBAMR/files/15402649/configagain.log)

**Reply by boyceg** [reply]
What is the error you are getting when you try to run `make install`?

**Reply by boyceg** [reply]
(I should have written: what is the _full_ error you are getting when you try to run `make install`?)

Note that not all versions of libMesh are compatible with all versions of PETSc. I don't know if this is the case, but I would not be surprised if libMesh 1.7.1 (released in 2022) is not compatible with PETSc 3.21 (released in 2024).

Note that autoibamr is set up to use versions of libMesh and PETSc that are compatible with each other.

**Reply by agiordan96** [reply]
Hi! It says
`../LIBMESH/src/solvers/petsc_dm_wrapper.C:191:18: error: no matching function for call to 'DMCreateSectionSubDM'
          ierr = DMCreateSectionSubDM(dm, numFields, fields, is, subdm);`

and at the end of the installation process, it goes

`4 warnings generated.
4 warnings generated.
4 warnings generated.
make: *** [install-recursive] Error 1`

**Reply by boyceg** [reply]
This means that your PETSc and libMesh versions are not compatible with each other. There are several options:
1. use libMesh GitHub `master` (note that you have to do more than just pull libMesh, you also need to set up libMesh submodules; [instructions are here](https://github.com/libMesh/libmesh?tab=readme-ov-file#build-instructions))
2. use a version of PETSc that predates the libMesh 1.7.1 release
3. use [`autoibamr`](https://github.com/IBAMR/autoibamr), which take care of libMesh and PETSc (and all of IBAMR's other dependencies) automatically

**Reply by agiordan96** [reply]
HI, I Was going to tell you just now that I went for libMesh master just now, you preceded me!

**Reply by agiordan96** [reply]
Since we ' re talking, 
1) thanks for your help!
2) Is there any tutorial / document for starters with IBAMR?

**Reply by agiordan96** [reply]
Hi Boyce, while installing an optimised build of IBAMR, I surprisingly get 1 warning and 20 errors. Configuration went smoothly, on the other hand. Attached my log file. Thanks!
[config.log](https://github.com/IBAMR/IBAMR/files/15422825/config.log)

**Reply by boyceg** [reply]
Can you also send the errors?

**Reply by agiordan96** [reply]
Oops!

[Console_output.pdf](https://github.com/IBAMR/IBAMR/files/15428711/Console_output.pdf)

**Reply by drwells** [reply]
It looks like the root-cause is that libMesh is set up with C++17 support but IBAMR is not. This is a bug in our autotools scripts.

For now, could you re-run `./configure` but add `-std=c++17` to CXXFLAGS? Alternatively, cmake correctly handles language version corner-cases much better than our autotools scripts - using that will also fix the problem.

**Reply by agiordan96** [reply]
Done! Lots of warnings!

[output2.pdf](https://github.com/IBAMR/IBAMR/files/15438046/output2.pdf)

**Reply by agiordan96** [reply]
Hi both,

I am able to configure IBAMR using c++17 (as @drwells suggested) and it configures smoothly. However, for the optimised version, it looks for the library libnetcdf.la in ,libmesh-directory / libmesh version number / version-opt / lib / libnetcdf.la, whereas I configured my lib mesh to have one less folder (the version number). On the other hand, this works fine with the debug version. While configuring the opt version of IBAMR, the correct lib mesh directories were given. Any thoughts on how to solve this? Thanks!
[ibamr_config_&_makeerror.pdf](https://github.com/IBAMR/IBAMR/files/15453527/ibamr_config_._makeerror.pdf)
[libmesh_opt_config.pdf](https://github.com/IBAMR/IBAMR/files/15453546/libmesh_opt_config.pdf)

**Reply by drwells** [reply]
Can you check that libMesh actually installed netcdf there? I don't think that that's a problem we can solve - i.e., if libMesh exports a specific list of dependencies but doesn't install them then there's no way for us to detect it.

**Reply by boyceg** [reply]
@agiordan96, it looks like you are configuring libMesh with `--prefix=/Users/andreagiordano/sim/libmesh-
again/1.7.1-opt` and configuring IBAMR with `--with-libmesh=/Users/andreagiordano/sim/libmesh-
again/1.7.1-opt`, but when you are building the examples, it looks like this directory has become `/ Users/andreagiordano/sim/libmesh-again/1.7.1/1.7.1-opt`, which does not seem consistent with your `config.log` file.

Can you try running `make V=1 examples` and report the full output.

Can you also try rebuilding IBAMR from scratch and make sure that you specify `/Users/andreagiordano/sim/libmesh-
again/1.7.1-opt` consistently?

Finally, would you be able to post plain text files instead of PDFs?

**Reply by agiordan96** [reply]
Hi there, 

thanks!

@boyceg Yes, that is what I was saying. I will try again, but I tried three times before writing you (where I re configured also lib mesh). Here you go with the attached output:
[v=1_Output.txt](https://github.com/IBAMR/IBAMR/files/15484665/v.1_Output.txt)

Please note that libmesh main folder  is called <libmesh-again>, and not <libmesh- again> (I am quite sure you could not determine this from the pdf). 

@drwells yes, libmesh installs netcfd in libmesh-again/1.7.1-opt/include/netcfd.h, whereas libnetcfd.la in found in libmesh-again/1.7.1-opt/lib/

**Reply by boyceg** [reply]
Please send the output of `make V=1 examples`, not `make v=1 examples`.

I will note again that the linker seems to be looking for these files in `/Users/andreagiordano/sim/libmesh-again/1.7.1/1.7.1-opt` instead of in `/Users/andreagiordano/sim/libmesh-again/1.7.1-opt`. I do not think there is anything in the IBAMR configuration system that can change directory names, and so I think that somehow you are specifying `/Users/andreagiordano/sim/libmesh-again/1.7.1/1.7.1-opt` as the directory where libMesh is installed.

Can you please also send the output of:
```
$ ls /Users/andreagiordano/sim/libmesh-again/1.7.1
$ ls /Users/andreagiordano/sim/libmesh-again/1.7.1/1.7.1-opt
$ ls /Users/andreagiordano/sim/libmesh-again/1.7.1-opt
```

**Reply by agiordan96** [reply]
@boyceg Sure! Here you go...
[output_ls.txt](https://github.com/IBAMR/IBAMR/files/15485139/output_ls.txt)
[output_V=1.txt](https://github.com/IBAMR/IBAMR/files/15485140/output_V.1.txt)

**Reply by boyceg** [reply]
OK, it looks like IBAMR thinks that libMesh is installed in `/Users/andreagiordano/sim/libmesh-again/1.7.1-opt`, which is consistent with your filesystem. However, it also looks like libMesh thinks that it is installed in `/Users/andreagiordano/sim/libmesh-again/1.7.1/1.7.1-opt`.

Can you send the contents of `/Users/andreagiordano/sim/libmesh-again/1.7.1-opt/lib/libmesh_opt.la`?

Is it possible that you initially installed libMesh in the directory `/Users/andreagiordano/sim/libmesh-again/1.7.1/1.7.1-opt` and then moved it to `/Users/andreagiordano/sim/libmesh-again/1.7.1-opt`? If so,  I would suggest:
1. remove the libMesh build directory
2. re-configure and re-compile libMesh; when you run libMesh `configure`, make sure that you set the flag `--prefix=/Users/andreagiordano/sim/libmesh-again/1.7.1-opt`
3. remove the IBAMR build directory & re-run IBAMR `configure` and `make`

**Reply by agiordan96** [reply]
Hi @boyceg, yes, it is very weird. Nope, no chance I did this. In libmesh_opt.la you ll find a reference to a folder 1.7.1/1.7.1-opt. Then, should I simply re build libmesh inserting the version number (1.7.1) in my file_system and solve the issue? Maybe it is a slight bug or a sort of requirement of libmesh, even if it sounds improbable. 
[libmesh_opt.txt](https://github.com/IBAMR/IBAMR/files/15485681/libmesh_opt.txt)

**Reply by boyceg** [reply]
You could try manually editing `libmesh_opt.la` to point to the correct directory. If you are able to reliably reproduce this issue, you might consider following up with the libMesh development team to see if there is an issue with their build system.

**Reply by agiordan96** [reply]
Hi, thank you all for your help. I built my directory structure as follows: libmesh/1.7.1/    etc. It worked. I will see if I can reproduce the issue. 
Bye!
